### PR TITLE
Add validation to HTLC Extend

### LINKED ISSUE #1696: Insufficient htlc_extend_operation validation
**Bug Description**
1. Everyone can extend a HTLC, which means funds inside can be locked forever (although need to pay per-day fee).
https://github.com/bitshares/bitshares-core/blob/11432f38a63416c3204278e6ea1d4f8133369585/libraries/chain/htlc_evaluator.cpp#L123-L127

2. Theoretically the sender can overflow a HTLC and got refunded (although may need to pay a large per-day fee).
https://github.com/bitshares/bitshares-core/blob/11432f38a63416c3204278e6ea1d4f8133369585/libraries/chain/htlc_evaluator.cpp#L129-L136

**Impacts**
Describe which portion(s) of BitShares Core may be impacted by this bug. Please tick at least one box.
- [ ] API (the application programming interface)
- [ ] Build (the build process or something prior to compiled code)
- [ ] CLI (the command line wallet)
- [ ] Deployment (the deployment process after building such as Docker, Travis, etc.)
- [ ] DEX (the Decentralized EXchange, market engine, etc.)
- [ ] P2P (the peer-to-peer network for transaction/block propagation)
- [ ] Performance (system or user efficiency, etc.)
- [x] Protocol (the blockchain logic, consensus, validation, etc.)
- [x] Security (the security of system or user data, etc.)
- [ ] UX (the User Experience)
- [ ] Other (please add below)

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Bug Report
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

### PR DESCRIPTION
**Pull request by jmjatlanta**
Fixes #1696

### CODE REVIEW

File: libraries/chain/htlc_evaluator.cpp [hunk 1]
```diff
@@ -123,6 +123,9 @@ namespace graphene {
       void_result htlc_extend_evaluator::do_evaluate(const htlc_extend_operation& o)
       {
          htlc_obj = &db().get<htlc_object>(o.htlc_id);
+         FC_ASSERT(o.update_issuer == htlc_obj->transfer.from, "HTLC may only be extended by its creator.");
+         FC_ASSERT(o.seconds_to_add <= fc::time_point_sec::maximum().sec_since_epoch()
+               - htlc_obj->conditions.time_lock.expiration.sec_since_epoch(), "Invalid number of seconds" );
```
Comment (abitmore) [inline libraries/chain/htlc_evaluator.cpp]: https://github.com/bitshares/bitshares-core/issues/1696#issuecomment-479834410:
>Validate: timeout_threshold < now() + GRAPHENE_HTLC_MAXIMUM_DURRATION

Comment (jmjatlanta) [reply libraries/chain/htlc_evaluator.cpp]: Good catch. Added that, plus an overflow check for fc::time_point_sec

File: libraries/chain/include/graphene/chain/protocol/htlc.hpp [hunk 2]
```diff
@@ -126,16 +126,16 @@ namespace graphene {
...
+               htlc_id(htlc_id), from(from), to(to), redeemer(redeemer), amount(amount) {}
 
          account_id_type fee_payer()const { return to; }
          void validate()const { FC_ASSERT( !"virtual operation" ); }
 
          share_type      calculate_fee(const fee_parameters_type& k)const { return 0; }
 
          htlc_id_type htlc_id;
-         account_id_type from, to;
+         account_id_type from, to, redeemer;
```
Comment (abitmore) [inline libraries/chain/include/graphene/chain/protocol/htlc.hpp]: I think we need to bump DB_VERSION if to add a field here.

Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: Done.

Comment (pmconrad) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: I disagree. We don't have any such operations in the database yet, therefore the change has no effect. We can avoid the replay.

Comment (abitmore) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: We can do the bump for testnet branch only.

Comment (jmjatlanta) [reply libraries/chain/include/graphene/chain/protocol/htlc.hpp]: Done

**Pull request merged**